apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RetryOptions.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>The type of retry to make.</summary>
+    public enum RetryType
+    {
+        /// <summary>Requests will not be retried.</summary>
+        None,
+
+        /// <summary>Requests will be retried.</summary>
+        Retry,
+    }
+
+    /// <summary>How to handle a full buffer holding items from calls to be retried.</summary>
+    public enum BufferOverflow
+    {
+        /// <summary>Throws and exception when the buffer is full.</summary>
+        ThrowException,
+
+        /// <summary>Ignore new entries.</summary>
+        IgnoreNewEntries
+    }
+
+    /// <summary>How to handle exceptions from calls.</summary>
+    public enum ExceptionHandling
+    {
+        /// <summary>
+        /// Let the errors propagate up after they have been retried if they continually fail.
+        /// This may kill the running process.
+        /// </summary>
+        Propagate,
+
+        /// <summary>Swallow and ignore the errors.</summary>
+        Swallow,
+    }
+
+    /// <summary>
+    /// Options that define retry logic for requests that failed.
+    /// </summary>
+    public sealed class RetryOptions
+    {
+        /// <summary>The default buffer size in bytes. 2^18 = 262144.</summary>
+        private const int DefaultBufferSize = 262144;
```

Comment:
For powers of 2, I usually like to define my constants by left shifting 1 by the exponent amount, like this:
```csharp
1 << 18
```
But that's just me. Feel free to ignore.